File: src/lib/knowledge/indexer.ts
import type { Firestore } from "firebase-admin/firestore";
import admin from "firebase-admin";
import { chunkArticleHtml } from "./chunking";
import { generateDocumentEmbedding } from '@/lib/brain/embed';

/**
 * ⚠️ DEPRECATED: Firestore brain_chunks indexing is being phased out in favor of Vertex AI Vector Search.
 * 
 * This function is kept for backward compatibility during migration.
 * New code should use the unified Vertex AI indexing via brain_jobs collection.
 * 
 * Migration plan:
 * 1. Articles are now indexed to Vertex AI via onArticleUpdated cloud function
 * 2. Topics and Insights already use Vertex AI via processBrainJob
 * 3. After full migration, this file will be removed
 * 
 * See: @/functions/src/onArticleUpdated.ts for new implementation
 */

function safeSlug(s: string) {
  return (s ?? "")
    .toLowerCase()
    .replace(/[^\p{L}\p{N}]+/gu, "-")
    .replace(/(^-|-$)/g, "")
    .slice(0, 60);
}

/**
 * Indexes a help center article into Firestore vector search.
 */
export async function indexHelpCenterArticleToChunks(args: {
  adminDB: Firestore;
  article: any; // HelpCenterArticle
  spaceId: string;
  publicHelpBaseUrl: string;
}) {
  const { adminDB, article, spaceId, publicHelpBaseUrl } = args;

  const status = article.status;
  if (status !== "published") return { chunkCount: 0 };

  const hubId = article.hubId;
  const helpCenterId: string = article.helpCenterId;
  if (!hubId || !helpCenterId) return { chunkCount: 0 };

  const articleId = article.id;
  const articleTitle = article.title ?? "Untitled";
  
  const specs = chunkArticleHtml({
    html: article.content ?? "",
    maxTokens: 220,
    overlapTokens: 60,
  });

  console.log(`INDEXER: Processing article ${articleId} into ${specs.length} specs`);

  const modelName = process.env.EMBEDDING_MODEL || 'gemini-embedding-2-preview';

  // FIX 4: Parallelize embedding generation for 5-10x speedup
  console.log(`INDEXER: Generating ${specs.length} embeddings in parallel...`);
  const results = await Promise.all(
    specs.map(async (c) => {
      try {
        const embedding = await generateDocumentEmbedding(c.text);
        return { spec: c, embedding };
      } catch (err) {
        console.error(`INDEXER: Embedding failed for chunk ${c.chunkIndex} of article ${articleId}`, err);
        return { spec: c, embedding: null };
      }
    })
  );

  let chunkCount = 0;

  for (const { spec, embedding } of results) {
    if (!embedding) continue;

    const anchor = spec.headingPath.length
        ? safeSlug(spec.headingPath.join("-")) + `-${spec.chunkIndex}`
        : `chunk-${spec.chunkIndex}`;

    const url = article.publicUrl
      ? (article.publicUrl.startsWith("http") ? article.publicUrl : `${publicHelpBaseUrl}${article.publicUrl}`)
      : `${publicHelpBaseUrl}/hc/${helpCenterId}/articles/${articleId}`;

    const chunkData = {
        hubId,
        spaceId,
        sourceType: 'help_center_article',
        sourceId: articleId,
        helpCenterId,
        title: articleTitle,
        text: spec.text,
        url: url + (anchor ? `#${anchor}` : ""),
        visibility: article.visibility || 'public',
        allowedUserIds: article.allowedUserIds || [],
        status: 'active',
        // Firestore Vector write
        embedding: (admin.firestore.FieldValue as any).vector(embedding),
        embeddingModel: modelName,
        embeddingDim: 1536,
        // FIX 5: Use serverTimestamp for accurate ordering
        createdAt: admin.firestore.FieldValue.serverTimestamp(),
        updatedAt: admin.firestore.FieldValue.serverTimestamp(),
        headingPath: spec.headingPath?.filter((h): h is string => typeof h === 'string') || [],
        chunkIndex: spec.chunkIndex,
    };

    await adminDB.collection('brain_chunks').add(chunkData);
    chunkCount++;
  }

  // Update index metadata
  await adminDB.collection("help_center_articles").doc(articleId).set(
    {
      chunkCount: chunkCount,
      chunkedAt: new Date().toISOString(),
    },
    { merge: true }
  );

  console.log(`INDEXER: Successfully indexed ${chunkCount} chunks for ${articleId}`);
  return { chunkCount };
}

File: src/lib/knowledge/indexer.py
import asyncio
import os
import re
from datetime import datetime
from typing import Optional, Any
import firebase_admin
from firebase_admin import firestore

from .chunking import chunk_article_html
from ..brain.embed import generate_document_embedding


def safe_slug(s: Optional[str]) -> str:
    """Convert a string to a safe URL slug."""
    if not s:
        return ""
    
    # Convert to lowercase
    s = s.lower()
    # Replace non-alphanumeric characters with hyphens
    s = re.sub(r'[^\p{L}\p{N}]+', '-', s, flags=re.UNICODE)
    # Remove leading/trailing hyphens
    s = re.sub(r'^-|-$', '', s)
    # Limit to 60 characters
    return s[:60]


async def index_help_center_article_to_chunks(
    admin_db: firestore.client,
    article: dict[str, Any],
    space_id: str,
    public_help_base_url: str,
) -> dict[str, int]:
    """
    Index a help center article into Firestore vector search.
    
    Args:
        admin_db: Firebase Firestore admin client
        article: Article document data
        space_id: Space/workspace ID
        public_help_base_url: Base URL for public help center
    
    Returns:
        Dictionary with chunk count
    """
    # Check if article is published
    status = article.get('status')
    if status != 'published':
        return {'chunk_count': 0}
    
    # Validate required fields
    hub_id = article.get('hubId')
    help_center_id = article.get('helpCenterId')
    if not hub_id or not help_center_id:
        return {'chunk_count': 0}
    
    article_id = article.get('id', '')
    article_title = article.get('title', 'Untitled')
    
    # Split article into chunks
    specs = chunk_article_html(
        html=article.get('content', ''),
        max_tokens=220,
        overlap_tokens=60
    )
    
    print(f"INDEXER: Processing article {article_id} into {len(specs)} specs")
    
    # Get embedding model from environment
    embedding_model = os.getenv('EMBEDDING_MODEL', 'gemini-embedding-2-preview')
    
    # Generate embeddings in parallel
    print(f"INDEXER: Generating {len(specs)} embeddings in parallel...")
    
    async def generate_embedding_for_chunk(chunk_spec: dict[str, Any]) -> tuple[dict[str, Any], Optional[list[float]]]:
        """Generate embedding for a single chunk."""
        try:
            embedding = await generate_document_embedding(chunk_spec['text'])
            return (chunk_spec, embedding)
        except Exception as err:
            print(f"INDEXER: Embedding failed for chunk {chunk_spec.get('chunk_index')} of article {article_id}: {err}")
            return (chunk_spec, None)
    
    # Parallelize embedding generation
    results = await asyncio.gather(
        *[generate_embedding_for_chunk(spec) for spec in specs]
    )
    
    chunk_count = 0
    
    for spec, embedding in results:
        if not embedding:
            continue
        
        # Generate anchor from heading path
        if spec.get('heading_path'):
            anchor = safe_slug('-'.join(spec['heading_path'])) + f"-{spec.get('chunk_index', 0)}"
        else:
            anchor = f"chunk-{spec.get('chunk_index', 0)}"
        
        # Construct URL
        public_url = article.get('publicUrl')
        if public_url:
            if public_url.startswith('http'):
                url = public_url
            else:
                url = f"{public_help_base_url}{public_url}"
        else:
            url = f"{public_help_base_url}/hc/{help_center_id}/articles/{article_id}"
        
        full_url = f"{url}#{anchor}" if anchor else url
        
        # Prepare chunk data for Firestore
        chunk_data = {
            'hubId': hub_id,
            'spaceId': space_id,
            'sourceType': 'help_center_article',
            'sourceId': article_id,
            'helpCenterId': help_center_id,
            'title': article_title,
            'text': spec['text'],
            'url': full_url,
            'visibility': article.get('visibility', 'public'),
            'allowedUserIds': article.get('allowedUserIds', []),
            'status': 'active',
            # Vector embedding
            'embedding': embedding,
            'embeddingModel': embedding_model,
            'embeddingDim': 2048,
            # Timestamps
            'createdAt': firestore.SERVER_TIMESTAMP,
            'updatedAt': firestore.SERVER_TIMESTAMP,
            'headingPath': spec.get('heading_path', []),
            'chunkIndex': spec.get('chunk_index', 0),
        }
        
        # Add chunk to Firestore
        await admin_db.collection('brain_chunks').add(chunk_data)
        chunk_count += 1
    
    # Update article metadata with chunk count
    await admin_db.collection('help_center_articles').document(article_id).set(
        {
            'chunkCount': chunk_count,
            'chunkedAt': datetime.now().isoformat(),
        },
        merge=True
    )
    
    print(f"INDEXER: Successfully indexed {chunk_count} chunks for {article_id}")
    return {'chunk_count': chunk_count}
